worlds/sc2/mission_order/mission_pools.py

```diff
@@ -102,6 +106,62 @@ def get_used_missions(self) -> List[SC2Mission]:
         """Returns a set of all missions used in the mission order."""
         return self._used_missions
 
+    def set_flag_balances(self, flag_ratios: Dict[MissionFlag, int], flag_weights: Dict[MissionFlag, int]):
+        # Ensure the ratios are percentages
+        ratio_sum = sum(ratio for ratio in flag_ratios.values())
+        self._flag_ratios = {flag: ratio / ratio_sum for flag, ratio in flag_ratios.items()}
+        self._flag_weights = flag_weights
+
+    def pick_balanced_mission(self, world: World, pool: List[int]) -> int:
+        """Applies ratio-based and weight-based balancing to pick a preferred mission from a given mission pool."""
+        # Currently only used for race balancing
+        # Untested for flags that may overlap or not be present at all, but should at least generate
+        balanced_pool = pool
+        if len(self._flag_ratios) > 0:
+            relevant_used_flag_count = max(sum(self._used_flags.get(flag, 0) for flag in self._flag_ratios), 1)
+            current_ratios = {
+                flag: self._used_flags.get(flag, 0) / relevant_used_flag_count
+                for flag in self._flag_ratios
+            }
+            # Desirability of missions is the difference between target and current ratios for relevant flags
+            flag_scores = {
+                flag: self._flag_ratios[flag] - current_ratios[flag]
+                for flag in self._flag_ratios
+            }
+            mission_scores = [
+                sum(
+                    flag_scores[flag] for flag in self._flag_ratios
+                    if flag in lookup_id_to_mission[mission].flags
+                )
+                for mission in balanced_pool
+            ]
+            # Only keep the missions that create the best balance
+            best_score = max(mission_scores)
+            balanced_pool = [mission for idx, mission in enumerate(balanced_pool) if mission_scores[idx] == best_score]
```

Comment:
Won't it cause some even alternation between mission races (if all races are enabled)? 
If I have Gauntlet with all races, all race mission pools are even, the races will alternate in this pattern:
Picked race, one of the unpicked races in previous pick, the remainder race. 
Getting 3 missions in a row is impossible, and 2nd, 5th, etc mission will have a distinct race from its neighbors (and each of those will be a different race too)

Like if I have WoL + Prophecy, it'll shift Prophecy missions to early placed missions

Reply:
Otherwise we could get the total size and calculate the distribution first. If we run out of eligible races, recalculate the remainder. That'd make the placement more random (as you can pick a random mission from random race as long as you're able to)

Reply:
It will cause that sort of alternation in principle, but there's some layers of randomness to the mission pick. Slots pick their missions not in order of depth, but in order of difficulty and randomly within each difficulty, so WoL + Prophecy will push the Prophecy missions to their earliest allowed difficulty, but not necessarily the earliest accessible slots.

Here's some sample generations, all size 25 Gauntlet, race swaps off.
WoL + Prophecy:
![python_FjVdoWg7D4](https://github.com/user-attachments/assets/198043fa-14a1-4001-8875-e4503eb0268b)

WoL + LotV:
![python_71DSc6oCpv](https://github.com/user-attachments/assets/8fde7d22-3ec7-416f-8389-96defddf4cfa)

WoL + LotV + HotS:
![python_lPeVslvLNp](https://github.com/user-attachments/assets/45bce167-2dd4-436b-b5b8-2502e77c2abe)

I think there's *some* bias towards very early and very late alternations because Starter and Very Hard difficulties have relatively few slots, but after looking at more generations (checking Gauntlet was a very good idea btw!) I'm not convinced it's much of a problem. There's definitely some psychological "I can see a pattern because I'm trying to find a pattern" thing going on as well, which makes it hard to judge from just a few generations. Might be better to just let people play with it and see if it comes up as an issue.

I also ran into this outlier, with a whole five consecutive Zerg missions at the end, which I just thought was cool.
![python_QS8eZIzAlE](https://github.com/user-attachments/assets/b1948a69-5bd7-4702-95a7-39f622bda100)

Reply:
The placement is more random than I've expected at the first sight